pkg/resourcemanager/controller/tokenrequestor/reconciler.go

```diff
@@ -0,0 +1,226 @@
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	expirationSeconds, err := tokenExpirationSeconds(secret)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	tokenRequest, err := r.createServiceAccountToken(ctx, serviceAccount, expirationSeconds)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	renewDuration := r.renewDuration(tokenRequest.Status.ExpirationTimestamp.Time)
+
+	if err := r.reconcileSecret(ctx, secret, tokenRequest.Status.Token, renewDuration); err != nil {
+		return reconcile.Result{}, fmt.Errorf("could not update Secret with token: %w", err)
+	}
+
+	return reconcile.Result{Requeue: true, RequeueAfter: renewDuration}, nil
+}
+
+func shouldDeleteServiceAccount(secret *corev1.Secret) bool {
+	return secret.Annotations[resourcesv1alpha1.ServiceAccountSkipDeletion] != "true"
+}
+
+func (r *reconciler) reconcileServiceAccount(ctx context.Context, secret *corev1.Secret) (*corev1.ServiceAccount, error) {
+	serviceAccount := getServiceAccountFromAnnotations(secret.Annotations)
+
+	if _, err := controllerutil.CreateOrUpdate(ctx, r.targetClient, serviceAccount, func() error {
+		serviceAccount.AutomountServiceAccountToken = pointer.Bool(false)
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return serviceAccount, nil
+}
+
+func (r *reconciler) deleteServiceAccount(ctx context.Context, secret *corev1.Secret) error {
+	serviceAccount := getServiceAccountFromAnnotations(secret.Annotations)
+
+	return client.IgnoreNotFound(r.targetClient.Delete(ctx, serviceAccount))
+}
+
+func (r *reconciler) reconcileSecret(ctx context.Context, secret *corev1.Secret, token string, renewDuration time.Duration) error {
+	patch := client.MergeFrom(secret.DeepCopy())
+
+	if secret.Data == nil {
+		secret.Data = make(map[string][]byte, 1)
+	}
+
+	secret.Data[resourcesv1alpha1.DataKeyToken] = []byte(token)
+	metav1.SetMetaDataAnnotation(&secret.ObjectMeta, resourcesv1alpha1.ServiceAccountTokenRenewTimestamp, r.clock.Now().UTC().Add(renewDuration).Format(layout))
+
+	return r.client.Patch(ctx, secret, patch)
+}
+
+func (r *reconciler) createServiceAccountToken(ctx context.Context, sa *corev1.ServiceAccount, expirationSeconds int64) (*authenticationv1.TokenRequest, error) {
+	tokenRequest := &authenticationv1.TokenRequest{
+		Spec: authenticationv1.TokenRequestSpec{
+			Audiences:         []string{v1beta1constants.GardenerAudience},
+			ExpirationSeconds: &expirationSeconds,
+		},
+	}
+
+	return r.targetCoreV1Client.ServiceAccounts(sa.Namespace).CreateToken(ctx, sa.Name, tokenRequest, metav1.CreateOptions{})
+}
+
+func (r *reconciler) requeue(renewTimestamp string) (bool, time.Duration, error) {
+	if len(renewTimestamp) == 0 {
+		return false, 0, nil
+	}
+
+	renewTime, err := time.Parse(layout, renewTimestamp)
+	if err != nil {
+		return false, 0, fmt.Errorf("could not parse renew timestamp: %w", err)
+	}
+
+	if r.clock.Now().UTC().Before(renewTime) {
+		return true, renewTime.Sub(r.clock.Now().UTC()), nil
```

Comment:
```suggestion
	if r.clock.Now().UTC().Before(renewTime.UTC()) {
		return true, renewTime.UTC().Sub(r.clock.Now().UTC()), nil
```
To be on the safe side?